### ISSUE TITLE
install() function not self-retriggering properly

### ISSUE THREAD

**Issue by penguin-wrangler** [derekantrican/GAS-ICS-Sync#137]
[This is with v5.2]

After manually running the install() function once, it successfully triggers the startSync() function to run a few moments later, but the trigger to re-run the install() function again after howFrequent minutes happens only once, and that instance does not trigger startSync() or install() again. The syncing happens properly the one time (good logs, no errors, entries properly synced). The install() function does not output any logs.

i.e. in the logs I see, i sequence, install() run manually (by Editor), then startSync() run by trigger, then howFrequent minutes later install() run by trigger. And then nothing else after that.

Every few days I come back and manually run the install() function again, and the same pattern of executions as described above happens again. The expected continuous re-triggering of install() is not happening.

**Comment by jonas0b1011001** [comment]
Please change [this](https://github.com/derekantrican/GAS-ICS-Sync/blob/115531fd1faab4b61b1dbab960e1d6367b21f285/Code.gs#L101) line to
`Logger.log(e);`
and check the log again once it failed.
If there is no output it will be really tough to debug.

This should not make a difference, but maybe try updating to 5.3

**Comment by penguin-wrangler** [comment]
Enabled logging of the exception of the catch{} block as you suggested, which resulted in nothing being logged, so I don't think an exception was being thrown. So I then added additional logging before and after each trigger add as follows:

function install(){
  try{
    //Delete any already existing triggers so we don't create excessive triggers
    deleteAllTriggers();
    
    if (howFrequent < 1){
      throw "[ERROR] \"howFrequent\" must be greater than 0.";
    }
    else{
      Logger.log("About to set install() trigger for now+25 minutes");
      ScriptApp.newTrigger("install").timeBased().after(howFrequent * 60 * 1000).create();//Schedule next Execution
      Logger.log("Done set install() trigger for now+25 minutes");
      Logger.log("About to set startSync() trigger for now+1 second");
      ScriptApp.newTrigger("startSync").timeBased().after(1000).create();//Start the sync routine
      Logger.log("Done set startSync() trigger for now+1 second");
    }
  }catch(e){
    Logger.log(e);
    //#install();//Retry on error
  }
}

The same pattern of executions resulted, before no further executions:

ICS-Syncer | install | Trigger | Jun 22, 2020, 3:34:45 PM | 32.393 s | Completed
ICS-Syncer | startSync | Trigger | Jun 22, 2020, 3:10:08 PM | 2.867 s | Completed
ICS-Syncer | install | Editor | Jun 22, 2020, 3:09:08 PM | 1.661 s | Completed


Both instances of install() logged the same info, like this:

Jun 22, 2020, 3:34:46 PM Info About to set install() trigger for now+25 minutes
Jun 22, 2020, 3:35:07 PM Info Done set install() trigger for now+25 minutes
Jun 22, 2020, 3:35:07 PM Info About to set startSync() trigger for now+1 second
Jun 22, 2020, 3:35:17 PM Info Done set startSync() trigger for now+1 second

Notice that sometimes the calls to create a new trigger take an exceptionally long time (21 seconds, and 10 seconds, respectively in this instance). I don't *think* that's relevant, but you may know better.

So even though the second instance of install() (triggered by the first manual run) seems to have successfully added two new time triggers (because there was no exception, and the "Done" logging happens), they don't ever actually fire.

**Comment by jonas0b1011001** [comment]
If you visit [https://script.google.com/home/triggers](https://script.google.com/home/triggers) do you see both entries for install() and startSync()?

Also, if you open the script in the editor, click on _Edit_ -> _Current Projects triggers_ you can manually add a trigger that runs startSync each x minutes. Does it work that way?

**Comment by penguin-wrangler** [comment]
The triggers appear in a variety of states depending when I look:

1) Immediately after running install() for the first time, there would be two time-based triggers (one to run startSync() and one to run install())
2) After the first triggered run of startSync(), it's trigger would show as disabled. (because it was a one-time event, that makes sense)
3) After 25 minutes, once the install() trigger had fired, that trigger would then also show as disabled (i.e. now there are two disabled triggers) This is the part that is odd, because running install() should remove all the triggers, and re-add new triggers for both install() and startSync(), but this doesn't appear to happen, even though the newTrigger.create() appears to succeed.

Perhaps the deleteAllTriggers() call at the start of install() isn't working, and then the subsequent re-adding of the triggers fails because it already exists? But then wouldn't that be an exception?

In the meantime, I did set up a manually recurring trigger schedule for startSync() as you suggested (set every 5 minutes for now, for testing), and it is repeatedly firing every 5 minutes.

I presume there was some logic behind repeatedly removing and re-adding one-time triggers rather than adding a recurring trigger once?

**Comment by jonas0b1011001** [comment]
> Perhaps the deleteAllTriggers() call at the start of install() isn't working

You can test that by running the uninstall function, all it does is delete the script's triggers by calling _deleteAllTriggers_. After uninstall, the [list](https://script.google.com/home/triggers) should be empty.

> and then the subsequent re-adding of the triggers fails because it already exists? But then wouldn't that be an exception?

It does not matter if a similar trigger already exists, we are deleting them just to not leave thousands of old triggers in the list.

> I presume there was some logic behind repeatedly removing and re-adding one-time triggers rather than adding a recurring trigger once?

We decided to use one time triggers as they allow the users to define any interval they want. Repeating triggers are limited to 1, 5, 15, 30... minutes.

> In the meantime, I did set up a manually recurring trigger schedule for startSync() as you suggested (set every 5 minutes for now, for testing), and it is repeatedly firing every 5 minutes.

That's good news. That way you can at least use the script.
I really don't know why the install routine is failing for you. You are the first one to report such an issue and the code seems to work just as it is supposed to be.

**Comment by penguin-wrangler** [comment]
Thanks. I can see from your code how to iterate over all the existing triggers, so I'll probably do some more debugging to log the triggers at every step and see where it goes wrong. I'll update when I get around to that. Thanks for your help.

**Comment by CompKingCanada** [comment]
I am also having the same problem....  I had an old version running which I've uninstalled, and created a new project.  I don't know if they are conflicting..

**Comment by jonas0b1011001** [comment]
I just realized that this seems to be a duplicate of #104

The issue mentioned in [this comment](https://github.com/derekantrican/GAS-ICS-Sync/issues/104#issuecomment-616547049) seems to be not fixed by google yet and for some reason only affects a fraction of users..

**Comment by ubartsch** [comment]
Hi, 

I had a similar problem and tried a few things.
To resolve this, I disabled the trigger for install (why do you have to reinstall every hour?) - 

And I changed the trigger generator for startSync to a different method: 
https://developers.google.com/apps-script/reference/script/clock-trigger-builder#everyminutesn

Here's what I have now, startSync seems to run fine at the set interval: 

``` function install(){
  try{
    //Delete any already existing triggers so we don't create excessive triggers
    deleteAllTriggers();
    
    if (howFrequent < 1){
      throw "[ERROR] \"howFrequent\" must be greater than 0.";
    }
    else{
      //ScriptApp.newTrigger("install").timeBased().everyHours(24).create();//Schedule install interval
      ScriptApp.newTrigger("startSync").timeBased().everyMinutes(15).create();//schedule sync interval
      startSync(); //immediate sync
      
    }
  }catch(e){
    install();//Retry on error
  }
}
```
I will now try to set the install trigger to a longer interval, prob every week:
https://developers.google.com/apps-script/reference/script/clock-trigger-builder#everyweeksn
Although I am still not sure why that is actually necessary?


Either way, I hope that helps,
U

**Comment by ubartsch** [comment]
New install function: 

```
function install(){
  try{
    //Delete any already existing triggers so we don't create excessive triggers
    deleteAllTriggers();
    
    if (howFrequent < 1){
      throw "[ERROR] \"howFrequent\" must be greater than 0.";
    }
    else{
      ScriptApp.newTrigger("install").timeBased().everyWeeks(1).create();//Schedule next Execution
      ScriptApp.newTrigger("startSync").timeBased().everyMinutes(howFrequent).create();//Start the sync routine
      startSync();
      
    }
  }catch(e){
    install();//Retry on error
  }
}
```

**Comment by jonas0b1011001** [comment]
> I will now try to set the install trigger to a longer interval, prob every week:
> https://developers.google.com/apps-script/reference/script/clock-trigger-builder#everyweeksn
> Although I am still not sure why that is actually necessary?

In your case it is not necessary, 
````
function install(){
      ScriptApp.newTrigger("startSync").timeBased().everyMinutes(howFrequent).create();
}
````
will work just fine.

See [this](https://github.com/derekantrican/GAS-ICS-Sync/issues/137#issuecomment-649793155) comment for details on why we use a different method.

**Comment by ubartsch** [comment]
I see, you are building new triggers on each run because of the time restrictions. 
I can live with every 15 mins, seems to use triggers like they were intended as well. 
Anyway, many thanks for putting this together in the first place, it's a very useful tool :)

**Comment by derekantrican** [comment]
@jonas0b1011001 I know that we originally added the `ScriptApp.newTrigger("startSync").timeBased().after(1000).create(); //Start the sync routine` as a sort of "async function". The new V8 runtime [supports async functions](https://developers.google.com/apps-script/guides/v8-runtime#improved_function_detection). I'm not very familiar with JS but if you are, feel free to convert to this functionality.

**Comment by mbloodk** [comment]
same issue, after some time triggers disappeared without any error.

**Comment by jonas0b1011001** [comment]
> @jonas0b1011001 I know that we originally added the `ScriptApp.newTrigger("startSync").timeBased().after(1000).create(); //Start the sync routine` as a sort of "async function". The new V8 runtime [supports async functions](https://developers.google.com/apps-script/guides/v8-runtime#improved_function_detection). I'm not very familiar with JS but if you are, feel free to convert to this functionality.

I have not looked into it but i don't think it will help fix the issue as the install function also seems to stop running eventually.

I guess we have two options:
1. Manually check if an active trigger exists for _install()_ and _startSync()_ at the end of install(), if not we can try to create them once again. This could work as it seems to be a temporary communication problem. The 300 seconds execution time _should_ be enough to eventually create the triggers successfully before the function times out.
2. Revert back to the old install method with time-based triggers. They do seem to run reliable.

Option 2 seems to be the easier and 100% reliable solution which is why i personally would go with that one.

**Comment by sunner** [comment]
I just manually deleted the install trigger and set the startSync trigger
be executed every 5 minutes. No code modification. So far so good.

**Comment by mastercko** [comment]
On this topic, I have been having the same issue as others have mentioned of this thread (install script will randomly stop re-triggering after a while and my calendar ends up stale) so I modified the script to use the `everyMinutes` trigger setting as others mentioned here. 

Additionally, to make the code more generally usable, I made a helper function to still use the `howFrequent` value as the user's intended amount of time but round it to the closest valid time setting usable for `everyMinutes` (1, 5 10, 15, 30). Hopefully this is a good middle ground to get the reliability of everyMinutes and also still capture the rough intent of how long a user wants to have the script wait between re-syncs.

In Code.gs, I modified `install`:
```
function install(){
  try{
    //Delete any already existing triggers so we don't create excessive triggers
    deleteAllTriggers();
    
    if (howFrequent < 1){
      throw "[ERROR] \"howFrequent\" must be greater than 0.";
    }
    else{
      ScriptApp.newTrigger("startSync").timeBased().everyMinutes(getValidTriggerFrequency(howFrequent)).create(); //Schedule sync routine to explicitly repeat
      ScriptApp.newTrigger("startSync").timeBased().after(1000).create();//Start the first sync routine immediately
    }
  }catch(e){
    install();//Retry on error
  }
}
```

and in Helpers.gs, I added `getValidTriggerFrequency`: 

```
function getValidTriggerFrequency(origFrequency) {
  if (origFrequency == null) 
    return 1; // if param is empty, just return 1, the lowest valid value.
  
  var adjFrequency = Math.round(origFrequency/5) * 5; // Set the number to be the closest divisible-by-5
  adjFrequency = Math.max(adjFrequency, 1); // Make sure the number is at least 1 (0 is not valid for the trigger)
  adjFrequency = Math.min(adjFrequency, 15); // Make sure the number is at most 15 (will check for the 30 value below)
  
  if((adjFrequency == 15) && (Math.abs(origFrequency-30) < Math.abs(origFrequency-15)))
    adjFrequency = 30; // If we adjusted to 15, but the original number is actually closer to 30, set it to 30 instead
  
  Logger.log("Intended frequency = "+origFrequency+", Adjusted frequency = "+adjFrequency);
  return adjFrequency;
}
```

I know that's probably not the most elegant way to do the calculation, but it works well enough for me.

**Comment by SusyV** [comment]
I been having the same problem in which my calendar(2) won't update unless I go to the script and manually do  "Run" > "Run function" > "startSync"
This is use to happen automatically.

**Comment by jonas0b1011001** [comment]
Can one of you please test if the following install routine is more stable:
```
function install(){
  //Delete any already existing triggers so we don't create excessive triggers
  deleteAllTriggers();
  
  if (howFrequent < 1)
    throw "[ERROR] \"howFrequent\" must be greater than 0.";
  
  var result = callWithBackoff(function(){
    ScriptApp.newTrigger("install").timeBased().after(howFrequent * 60 * 1000).create();//Schedule next Execution
    
    var triggers = ScriptApp.getProjectTriggers().filter(function(trigger){
      return trigger.getHandlerFunction() == "install";
    });
    
    if (triggers.length < 1)
      throw "[ERROR] Failed to create trigger for next execution.";
    
    return true;
  }, 10);
  
  if (result){
    ScriptApp.newTrigger("startSync").timeBased().after(1000).create();//Start the sync routine
  }
  else{
    throw "[ERROR] Installing the script failed."
  }
}
```

**Comment by SusyV** [comment]
You want me to replace the following with what you wrote above?:
function install(){
  try{
    //Delete any already existing triggers so we don't create excessive triggers
    deleteAllTriggers();
    
    if (howFrequent < 1){
      throw "[ERROR] \"howFrequent\" must be greater than 0.";
    }
    else{
      ScriptApp.newTrigger("install").timeBased().after(howFrequent * 60 * 1000).create();//Schedule next Execution
      ScriptApp.newTrigger("startSync").timeBased().after(1000).create();//Start the sync routine
    }
  }catch(e){
    install();//Retry on error
  }
}

**Comment by jonas0b1011001** [comment]
Yes, replace the code then run 'install' as usual and see if it fails again.

**Comment by mastercko** [comment]
I have updated my copy to use this install version and will let you know what I see over the next couple of days.

**Comment by cameo69** [comment]
> Can one of you please test if the following install routine is more stable:

Will try this.

**Comment by mastercko** [comment]
So for the past 3 days, it seems that this one has been working OK. Will check in again later this week.

**Comment by cameo69** [comment]
> > Can one of you please test if the following install routine is more stable:
> 
> Will try this.

So far I've seen >100 successful runs... Will recheck later this week.

**Comment by cameo69** [comment]
> > > Can one of you please test if the following install routine is more stable:
> > 
> > 
> > Will try this.
> 
> So far I've seen >100 successful runs... Will recheck later this week.

The updates failed yesterday evening. Please find a screenshot to the error message here: https://www.evernote.com/l/AHr7AepwjN9Hur_sSKizbrk3dhMqaaOTxjI

**Comment by mastercko** [comment]
I have also seen 5 errors in two days in startSync, similar to what @cameo69 is seeing I believe, though I unfortunately don't know German (apologies if I have that wrong!), so I can't tell if they are are exactly the same. 

> Sep 21, 2020, 10:23:08 AM	Error	Exception: API call to calendar.settings.get failed with error: Empty response at startSync(Code:226:42)
Sep 21, 2020, 12:12:52 PM	Error	Exception: API call to calendar.settings.get failed with error: Empty response at startSync(Code:226:42)
Sep 22, 2020, 11:31:45 AM	Error	We're sorry, there was an unexpected error while preparing top level objects.
Sep 22, 2020, 11:46:01 AM	Error	Exception: API call to calendar.settings.get failed with error: Empty response at startSync(Code:226:42)
Sep 22, 2020, 11:55:43 AM	Error	Exception: API call to calendar.events.list failed with error: Empty response at startSync(Code:199:39)

In my version of the script (modified with your updates and a few of my own changes), the lines referred to in the errors are:
**Code:199:**
`      var eventList = Calendar.Events.list(targetCalendarId, {showDeleted: false, privateExtendedProperty: "fromGAS=true", maxResults: 2500});`
**Code:226:**
`      var calendarTz = Calendar.Settings.get("timezone").value;`

@cameo69 where are you located? I wonder if we were seeing similar errors at similar times (I'm in California on Pacific Daylight Time). This might indicate that these errors are happening due to something on the Google side.

In any case, I assume the issues that I'm seeing are related to something different from the install issue in this issue report. I haven't seen any errors in `install()` runs at all so far. I have mine set for 5 minute updates, so I'm getting close to 300 runs of the `install()` script per day, all of which have worked. So as far as I can tell this new install script is continuing to work as expected.

**Comment by cameo69** [comment]
> @cameo69 where are you located? I wonder if we were seeing similar errors at similar times (I'm in California on Pacific Daylight Time). This might indicate that these errors are happening due to something on the Google side.

I am located in Germany, but I checked and my Google Account is set to English. Errors for other scripts I receive all in English?!?
Anyway, we have CEST which is UTC+2.

1. The first error is fairly unspecific, _"Beim Vorbereiten der übergeordneten Objekte ist ein unerwarteter Fehler aufgetreten."_ basically means that _preparing some parent object triggered an unexpected error_.

2. The 2nd error occured with call stack:
Helpers.gs: setupTargetCalendar: `var targetCalendar = Calendar.CalendarList.list().items.filter(function(cal) {`
Code.gs: StartSync(): `var targetCalendar = setupTargetCalendar(targetCalendarName);`



When I looked more closely I found some more older errors:

3. Some server error:

> Sep 22, 2020, 4:54:52 AM	Error	Exception: Ein Serverfehler ist aufgetreten. Bitte versuchen Sie es später erneut.
>     at install(Code:129:63)

Code.gs: install(): `ScriptApp.newTrigger("startSync").timeBased().after(1000).create();//Start the sync routine`

4. API-call failed

> Sep 21, 2020, 8:49:00 PM	Error	Exception: Beim API-Aufruf von "calendar.events.list" ist dieser Fehler aufgetreten: Antwort ist leer
>     at startSync(Code:197:39)

Code.gs: StartSync(): `var eventList = Calendar.Events.list(targetCalendarId, {showDeleted: false, privateExtendedProperty: "fromGAS=true", maxResults: 2500});`


hope that helps

**Comment by cameo69** [comment]
> 3. Some server error:
> 
> > Sep 22, 2020, 4:54:52 AM	Error	Exception: Ein Serverfehler ist aufgetreten. Bitte versuchen Sie es später erneut.
> > at install(Code:129:63)
> 
> Code.gs: install(): `ScriptApp.newTrigger("startSync").timeBased().after(1000).create();//Start the sync routine`

Found two more of those, but somehow it recovered and both triggers (Install and StartSync) are still in place.

**Comment by jonas0b1011001** [comment]
> > Code.gs: install(): `ScriptApp.newTrigger("startSync").timeBased().after(1000).create();//Start the sync routine`
> 
> Found two more of those, but somehow it recovered and both triggers (Install and StartSync) are still in place.

The error happened after the next run was scheduled, therefore it continued running.

Most of the errors you both shared are failed api calls which can not be fixed by us but we can detected them and just repeat the call till it succeeds.

> The updates failed yesterday evening. Please find a screenshot to the error message here: https://www.evernote.com/l/AHr7AepwjN9Hur_sSKizbrk3dhMqaaOTxjI

That error message is so vague though..i don't know if there's a way to fix or work around this from our side.

@derekantrican I think we should go back to the interval triggers, seems like new errors randomly appear each time we try a new workaround.

**Comment by derekantrican** [comment]
Sounds good to me. Should we just merge in #152 ?

**Comment by jonas0b1011001** [comment]
Sure, i have not had time to check the code though. If you did you can of course go ahead and approve or suggest changes to the PR.
